core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/value/decorator/IterableFromEntityPropertyValueSelector.java

```diff
@@ -0,0 +1,120 @@
+        implements IterableValueSelector<Solution_> {
+
+    private final FromEntityPropertyValueSelector<Solution_> childValueSelector;
+    private final boolean randomSelection;
+    private final FromEntityPropertyValueRangeDescriptor<Solution_> valueRangeDescriptor;
+    private InnerScoreDirector<Solution_, ?> innerScoreDirector = null;
+
+    public IterableFromEntityPropertyValueSelector(FromEntityPropertyValueSelector<Solution_> childValueSelector,
+            boolean randomSelection) {
+        this.childValueSelector = childValueSelector;
+        this.randomSelection = randomSelection;
+        this.valueRangeDescriptor = (FromEntityPropertyValueRangeDescriptor<Solution_>) childValueSelector
+                .getVariableDescriptor().getValueRangeDescriptor();
+    }
+
+    // ************************************************************************
+    // Life-cycle methods
+    // ************************************************************************
+
+    @Override
+    public void solvingStarted(SolverScope<Solution_> solverScope) {
+        super.solvingStarted(solverScope);
+        this.childValueSelector.solvingStarted(solverScope);
+        this.innerScoreDirector = solverScope.getScoreDirector();
+    }
+
+    @Override
+    public void solvingEnded(SolverScope<Solution_> solverScope) {
+        super.solvingEnded(solverScope);
+        childValueSelector.solvingEnded(solverScope);
+        this.innerScoreDirector = null;
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+    @Override
+    public GenuineVariableDescriptor<Solution_> getVariableDescriptor() {
+        return childValueSelector.getVariableDescriptor();
+    }
+
+    @Override
+    public long getSize(Object entity) {
+        return childValueSelector.getSize(entity);
+    }
+
+    @Override
+    public Iterator<Object> iterator(Object entity) {
+        return childValueSelector.iterator(entity);
+    }
+
+    @Override
+    public Iterator<Object> endingIterator(Object entity) {
+        return childValueSelector.endingIterator(entity);
+    }
+
+    @Override
+    public boolean isCountable() {
+        return valueRangeDescriptor.isCountable();
+    }
+
+    @Override
+    public boolean isNeverEnding() {
+        return randomSelection || !isCountable();
+    }
+
+    @Override
+    public long getSize() {
+        return innerScoreDirector.getValueRangeResolver().extractValueRangeSize(valueRangeDescriptor,
+                innerScoreDirector.getWorkingSolution(), null);
+    }
+
+    @Override
+    public Iterator<Object> iterator() {
+        var valueRange = (CountableValueRange<Object>) innerScoreDirector.getValueRangeResolver()
+                .extractValueRange(valueRangeDescriptor, innerScoreDirector.getWorkingSolution(), null);
```

Comment:
How can this be from entity, if it doesn't use any entity?

Reply:
The value change resolver will invoke the descriptor `FromEntityPropertyValueRangeDescriptor`, which uses the list of entities from the solution to generate a unique list of values. So, it does use the entity.